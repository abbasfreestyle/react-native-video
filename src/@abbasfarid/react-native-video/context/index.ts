export { useVideo, VideoProvider } from './video';
