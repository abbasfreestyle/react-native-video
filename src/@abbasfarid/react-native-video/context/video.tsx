import React, { createContext, ReactNode, useContext, useState } from 'react';

type IVideoProvider = {
  display: boolean;
  play: boolean;
  toggleDisplay: () => void;
  togglePlay: () => void;
};

const noop = () => {
  // do nothing;
};

/* istanbul ignore next */
const initialState: IVideoProvider = {
  display: false,
  play: false,
  toggleDisplay: noop,
  togglePlay: noop,
};

const VideoContext = createContext<IVideoProvider>(initialState);

export const VideoProvider = ({ children }: { children: ReactNode }) => {
  const value = useVideoProvider();
  return (
    <VideoContext.Provider value={value}>{children}</VideoContext.Provider>
  );
};

const useVideoProvider = () => {
  const [play, setPlay] = useState(false);
  const [display, setDisplay] = useState(false);

  const togglePlay = () => {
    setPlay((prev) => !prev);
  };

  const toggleDisplay = () => {
    setDisplay((prev) => !prev);
  };

  return {
    display,
    play,
    togglePlay,
    toggleDisplay,
  };
};

export const useVideo = () => useContext(VideoContext);
