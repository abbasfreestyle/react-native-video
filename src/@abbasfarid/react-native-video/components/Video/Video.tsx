import React, { useState } from 'react';
import { StyleSheet, View } from 'react-native';

import RNVideo, { VideoProperties } from 'react-native-video';

import { useVideo } from '../../context';
import { Displayer } from '..';

export type Props = {
  children?: React.ReactNode;
  source: VideoProperties['source'];
};

const styles = StyleSheet.create({
  container: {
    backgroundColor: 'black',
  },
  video: {
    alignSelf: 'stretch',
  },
  overlay: {
    position: 'absolute',
    top: 0,
    bottom: 0,
    right: 0,
    left: 0,
  },
});

export const VideoPlayer = ({ children, source }: Props) => {
  const [height, setHeight] = useState(0);
  const { play } = useVideo();
  return (
    <View
      onLayout={({
        nativeEvent: {
          layout: { width },
        },
      }) => {
        setHeight(width * 0.5625);
      }}
      style={styles.container}
      testID="video-container"
    >
      <RNVideo
        paused={!play}
        source={source}
        style={[{ height }, styles.video]}
        testID="rn-video"
      />
      <View style={styles.overlay}>
        <Displayer>{children}</Displayer>
      </View>
    </View>
  );
};
