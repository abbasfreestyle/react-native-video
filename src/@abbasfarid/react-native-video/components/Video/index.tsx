import React from 'react';

import { VideoProvider } from '../../context';
import { Props, VideoPlayer } from './Video';

export const Video = ({ children, ...videoProps }: Props) => (
  <VideoProvider>
    <VideoPlayer {...videoProps}>{children}</VideoPlayer>
  </VideoProvider>
);

export { Props as VideoPlayerProps } from './Video';
