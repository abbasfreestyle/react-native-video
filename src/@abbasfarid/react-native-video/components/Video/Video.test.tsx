import React from 'react';

import { act, fireEvent, render } from '@testing-library/react-native';

import { VideoProvider } from '../../context';
import { VideoPlayer } from './Video';

const renderComponent = () =>
  render(<VideoPlayer source={{ uri: 'mock-uri' }} />, {
    wrapper: VideoProvider,
  });

test('should match the snapshot', () => {
  const { toJSON } = renderComponent();

  expect(toJSON()).toMatchSnapshot();
});

test('should measure the height of the video', () => {
  const { getByTestId } = renderComponent();
  const ratio = 0.5625;
  const width = 300;
  const container = getByTestId('video-container');
  const video = getByTestId('rn-video');

  act(() => {
    fireEvent(container, 'layout', {
      nativeEvent: { layout: { width } },
    });
  });

  expect(video.props.style[0].height).toBe(width * ratio);
});
