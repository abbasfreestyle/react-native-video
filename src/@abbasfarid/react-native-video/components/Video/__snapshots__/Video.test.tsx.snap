// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "backgroundColor": "black",
    }
  }
  testID="video-container"
>
  <View
    paused={true}
    source={
      Object {
        "uri": "mock-uri",
      }
    }
    style={
      Array [
        Object {
          "height": 0,
        },
        Object {
          "alignSelf": "stretch",
        },
      ]
    }
    testID="rn-video"
  />
  <View
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <View
        animatedStyle={
          Object {
            "value": Object {
              "opacity": 0,
            },
          }
        }
        collapsable={false}
        style={
          Object {
            "backgroundColor": "rgba(0,0,0,0.25)",
            "height": "100%",
            "opacity": 0,
          }
        }
      />
    </View>
  </View>
</View>
`;
